text/0000-expand-impl-trait.md

```diff
@@ -0,0 +1,853 @@
+of `impl Trait`, there's been a desire to repurpose that syntax, and
+[instead write `Box<dyn Trait>`](https://github.com/rust-lang/rfcs/pull/1603) or
+some such to designate trait objects.
+
+That would, in particular, allow syntax like the following when taking a closure:
+
+```rust
+fn map<U>(self, f: FnOnce(T) -> U) -> Option<U>
+```
+
+The pros, cons, and logistics of such a change are out of scope for this
+RFC. However, it's taken as an assumption that we want to keep the door open to
+such a syntax, and so shouldn't stabilize any variant of `impl Trait` that lacks
+a good story for evolving into a bare `Trait` syntax later on.
+
+## Sticking with the `impl Trait` syntax
+
+This RFC proposes to stabilize the `impl Trait` feature with its current syntax,
+while also expanding it to encompass argument position. That means, in
+particular, *not* introducing an explicit `some`/`any` distinction.
+
+This choice is based partly on the core assumptions:
+
+- Assumption 1, we'll have a fully expressive syntax later.
+- Assumption 4, we can use the `some` semantics in return position and `any` in argument position, and almost always be right.
+- Assumption 5, we may want bare `Trait` syntax, which would not give "syntactic space" for a `some`/`any` distinction.
+
+One important question is: will people find it easier to understand and use
+`impl Trait`, or something like `some Trait` and `any Trait`? Having an explicit
+split may make it easier to understand what's going on. But on the other hand,
+it's a somewhat complicated distinction to make, and while you usually know
+*intuitively* what you want, being forced to spell it out by choosing the
+correct choice of `some` or `any` seems like an unnecessary burden, especially
+if the choice is almost always dictated by the position.
+
+Pedagogically, if we have an explicit syntax, we retain the option of
+explaining what's going on with `impl Trait` by "desugaring" it into that
+syntax. From that standpoint, `impl Trait` is meant purely for ergonomics, which
+means
+[not just what you type, but also what you have to remember](https://blog.rust-lang.org/2017/03/02/lang-ergonomics.html). Having
+`impl Trait` "just do the right thing" seems pretty clearly to be the right
+choice ergonomically.
+
+## Expansion to arguments
+
+This RFC proposes to allow `impl Trait` in function arguments, in addition to
+return position, with the `any Trait` semantics (as per assumption 4). In other
+words:
+
+```rust
+// These two are equivalent
+fn map<U>(self, f: impl FnOnce(T) -> U) -> Option<U>
+fn map<U, F>(self, f: F) -> Option<U> where F: FnOnce(T) -> U
+```
+
+However, this RFC also proposes to *disallow* use of `impl Trait` within `Fn`
+trait sugar or higher-ranked bounds, i.e. to disallow examples like the following:
+
+```rust
+fn foo(f: impl Fn(impl SomeTrait) -> impl OtherTrait)
+fn bar() -> (impl Fn(impl SomeTrait) -> impl OtherTrait)
+```
```

Comment:
Are you *also* planning to disallow of higher-order function pointers (*not* `Fn` traits) like this:

```rust
fn foo(g: fn (impl SomeTrait) -> impl Other Trait)
```

----

(I can *imagine* one possible semantics for this, where the types for the `impl`'s will be fixed at the point where `foo` is invoked; i.e. they would both be universally quantified at foo, like:
```rust
fn foo<S,O>(g: fn(S) -> O) where S: SomeTrait, O: OtherTrait
```
but it seems like it might be better to just disallow the syntax entirely...)